### ISSUE TITLE
Security vulnerability in version 5.1.0

### ISSUE THREAD

**Issue by abadfish** [thoughtbot/paperclip#2530]
Just got this message from GitHub: 
We found a potential security vulnerability in one of your dependencies.
The paperclip dependency defined in Gemfile.lock has a known high severity security vulnerability in version range >= 3.1.4, <= 5.1.0 and should be updated. 
I am using the latest version of paperclip. Is there a fix out there in the works? New version?

**Comment by guihatano** [comment]
I think this https://github.com/thoughtbot/paperclip/pull/2435 should work
But it needs to be merged yet, we will need to wait

**Comment by robbkidd** [comment]
Until a new version of the gem is released, would adding the following to an initializer be awful?

```ruby
internal_adapter_registry = Paperclip.io_adapters.instance_variable_get(:@registered_handlers)
worrisome_adapters = [Paperclip::UriAdapter, Paperclip::HttpUrlProxyAdapter, Paperclip::DataUriAdapter]

worrisome_adapters.each do |klass_to_remove|
  internal_adapter_registry.reject! { |_, klass| klass == klass_to_remove }
end
```

Results in:

```
> Paperclip.io_adapters.registered_handlers.map(&:second)
=> [Paperclip::EmptyStringAdapter,
 Paperclip::IdentityAdapter: ,
 Paperclip::FileAdapter,
 Paperclip::StringioAdapter,
 Paperclip::NilAdapter,
 Paperclip::AttachmentAdapter,
 Paperclip::UploadedFileAdapter]
```

**Comment by blimmer** [comment]
Agreed - a writeup in this repo would be really helpful now that GitHub is warning everyone. It's hard to tell who is affected, or the recommended workaround.

**Comment by edslocomb** [comment]
@jhulford I've been digging around for that myself; the description of the current vulnerability is apparently not public yet: https://hackerone.com/reports/209430

@guihatano that PR disables the Uri adapter by default, *but* users who still want the feature and re-enable it will presumably still be vulnerable, so I'm not convinced that the patch closes the security hole (whatever it might actually be, see above).

**Comment by edslocomb** [comment]
@reedloden that report is 4 years old, and a patch of some sort was merged 4 years ago.  Are you saying the fix missed a particular category of attack?  What do you mean by "though the reporter wasn't aware of it at the time?"

**Comment by edslocomb** [comment]
@reedloden thanks for the history, that does clear things up!

From the sound of it, users who *do* enable upload via URI will still be vulnerable after #2435 is merged.

**Comment by reedloden** [comment]
Perhaps incorporating https://github.com/jtdowney/private_address_check would be helpful towards a complete fix?

**Comment by paul-at** [comment]
@reedloden thank you for the detailed explanation!

My perception has always been that this is a widely known feature (even if undocumented) rather than a vulnerability. I first found it when one of the integrations to my app used it to copy attachments from one cloud to another (managed by paperclip).

**Comment by guihatano** [comment]
Hope they fix it soon

**Comment by mike-burns** [comment]
Fixing in #2531.

**Comment by mike-burns** [comment]
We have merged and released the fix from #2435 as Paperclip 5.2.0. Please upgrade as you see fit.

**Comment by rongutierrez** [comment]
I've released a blog post providing some more details on the vulnerability and some recommendations for handling uri_adapter validation when an app is using it - https://medium.com/in-the-weeds/all-about-paperclips-cve-2017-0889-server-side-request-forgery-ssrf-vulnerability-8cb2b1c96fe8

**Comment by dbelling** [comment]
Thanks for fixing this.

**Comment by pouellet** [comment]
Sorry for hijacking this closed thread, but it seems the best place to ask without reopening a new issue. 

The fix for this issue resulted in the removal of `DataUriAdapter` from the default list of registered adapters. After having a quick look at the code, I do not understand how this adapter could be vulnerable to SSRF. It is not a subclass of the `UriAdapter` class and, unless I'm missing something, doesn't do any network requests:

```
def extract_target(uri)
  data_uri_parts = uri.match(REGEXP) || []
  StringIO.new(Base64.decode64(data_uri_parts[2] || ""))
end
```

It's debatable wether it should be part of the default registered adapters, but I think it might be good to clarify in the documentation if it is actually vulnerable to SSFR.

**Comment by mike-burns** [comment]
@pouellet you are right, a data URI is not vulnerable to SSRF attack. It can pose other issues (this is a fun one from today, unrelated to Paperclip: https://thatoddmailbox.github.io/2017/01/28/iotaseed.html ), but that is unrelated to this specific CVE.
